### PR TITLE
Replace a vec of vecs in address_tables with a custom Array2d

### PR DESCRIPTION
**Pull request by ImplOfAnImpl**
I couldn't find an existing implementation of something like Array2d in a well-established 3rd-party crate, so I rolled my own.

### CODE REVIEW

**Review by iljakuklic — APPROVED** [review]

**Comment by iljakuklic** [comment]
I'd say https://crates.io/crates/ndarray is the canonical crate for n-dimensional arrays. Probably a bit of overkill for this use case though.

File: utils/src/array_2d.rs [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+        &mut self.array[start..end]
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn test() {
```
Comment (TheQuantumPhysicist) [inline utils/src/array_2d.rs:77]: Please add panic tests for out-of-range indexes

**Review by TheQuantumPhysicist — APPROVED** [review]

**Review by azarovh — APPROVED** [review]

**Comment by iljakuklic** [comment]
Anything still blocking this from merging?

**Comment by ImplOfAnImpl** [comment]
> Anything still blocking this from merging?

Nope

**Pull request merged**